fix(mme): fixing Pipelined managed tbl0 error when mme send UE info t…

Balaji-Jatla opened the pull request:
fix(mme): fixing Pipelined managed tbl0 error when mme send UE info to pipelined


Signed-off-by: Balaji-Jatla <sai.balaji@wavelabs.ai>

## Summary
Fixing Pipelined managed tbl0 error when mme send UE info to pipelined


## Test Plan
Tested with SRS-RAN simulator

## Additional Information
corresponding Zen hub Task id is #10943

Balaji-Jatla commented:
```
Jan 7 12:12:07 magma-dev-focal pipelined[505047]: UESessionSet {
Jan 7 12:12:07 magma-dev-focal pipelined[505047]:        subscriber_id {
Jan 7 12:12:07 magma-dev-focal pipelined[505047]:            id: "IMSI001010000000001"
Jan 7 12:12:07 magma-dev-focal pipelined[505047]:        }
Jan 7 12:12:07 magma-dev-focal pipelined[505047]:       precedence: 65535
Jan 7 12:12:07 magma-dev-focal pipelined[505047]:       ue_ipv4_address {
Jan 7 12:12:07 magma-dev-focal pipelined[505047]:              address: "\001\002\003\004"
Jan 7 12:12:07 magma-dev-focal pipelined[505047]:        }
Jan 7 12:12:07 magma-dev-focal pipelined[505047]:       enb_ip_address {
Jan 7 12:12:07 magma-dev-focal pipelined[505047]:           address: "\300\250<\241"
Jan 7 12:12:07 magma-dev-focal pipelined[505047]:       }
Jan 7 12:12:07 magma-dev-focal pipelined[505047]:       apn: "internet"
Jan 7 12:12:07 magma-dev-focal pipelined[505047]:        in_teid: 19
Jan 7 12:12:07 magma-dev-focal pipelined[505047]:        out_teid: 2
Jan 7 12:12:07 magma-dev-focal pipelined[505047]:        ue_session_state {
Jan 7 12:12:07 magma-dev-focal pipelined[505047]:        }
Jan 7 12:12:07 magma-dev-focal pipelined[505047]: }
```

pshelar reviewed (changes requested):
Can you also add S1AP integ test to avoid future regression.

koolzz commented:
If I remember correctly its not trivial to change a spgw.yml for only one integ test, @ardzoht is this true or am i making things up.

ardzoht commented:
> If I remember correctly its not trivial to change a spgw.yml for only one integ test, @ardzoht is this true or am i making things up.

This should be doable, you can do something similar as this https://sourcegraph.com/github.com/magma/magma/-/blob/lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py?L963-987 and then call that function on whichever integ test you need

koolzz commented:
> > If I remember correctly its not trivial to change a spgw.yml for only one integ test, @ardzoht is this true or am i making things up.
> 
> This should be doable, you can do something similar as this https://sourcegraph.com/github.com/magma/magma/-/blob/lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py?L963-987 and then call that function on whichever integ test you need

Good stuff thanks! @Balaji-Jatla could you add this

koolzz commented:
I've pulled the changes and built a new mme version and it seems like the issue is still present
```
Jan 20 01:03:47 phy-agw1 pipelined[2673709]: Traceback (most recent call last):
Jan 20 01:03:47 phy-agw1 pipelined[2673709]:   File "/usr/local/lib/python3.8/dist-packages/grpc/_common.py", line 86, in _transform
Jan 20 01:03:47 phy-agw1 pipelined[2673709]:     return transformer(message)
Jan 20 01:03:47 phy-agw1 pipelined[2673709]:   File "/usr/local/lib/python3.8/dist-packages/google/protobuf/internal/python_message.py", line 1093, in SerializeToString
Jan 20 01:03:47 phy-agw1 pipelined[2673709]:     if not self.IsInitialized():
Jan 20 01:03:47 phy-agw1 pipelined[2673709]: AttributeError: 'NoneType' object has no attribute 'IsInitialized'
```

Balaji-Jatla commented:
> Can you also add S1AP integ test to avoid future regression.

Hi @pshelar please find below file for S1AP test cases result.
[s1ap test cases result.txt](https://github.com/magma/magma/files/7910551/s1ap.test.cases.result.txt)

pshelar commented:
> > Can you also add S1AP integ test to avoid future regression.
> 
> Hi @pshelar please find below file for S1AP test cases result. [s1ap test cases result.txt](https://github.com/magma/magma/files/7910551/s1ap.test.cases.result.txt)

@Balaji-Jatla we need to add automated integ-test for pipelined-managed tbl0 mode. lets sync-up on slack if you have any questions.

Balaji-Jatla commented:
> > > Can you also add S1AP integ test to avoid future regression.
> > 
> > 
> > Hi @pshelar please find below file for S1AP test cases result. [s1ap test cases result.txt](https://github.com/magma/magma/files/7910551/s1ap.test.cases.result.txt)
> 
> @Balaji-Jatla we need to add automated integ-test for pipelined-managed tbl0 mode. lets sync-up on slack if you have any questions.

Sure @pshelar , I will do that.

koolzz commented on lte/protos/pipelined.proto:
```diff
@@ -449,7 +449,7 @@ message UESessionSet {
     uint32 out_teid = 9;
     UESessionState ue_session_state = 10;
-    IPFlowDL ip_flow_dl = 13;
+    IPFlowDL ip_flow_dl = 11;
```
not sure why we're bumping this number

Balaji-Jatla replied:
UESessionState ue_session_state = 10;  as using as a sequence number of gRPC attributes. so, we modified it as a 11.

koolzz replied:
Changing proto values is usually not safe, I suppose as we never run this code in production anywhere we can let it slide

Balaji-Jatla replied:
Wanted make  it sequential, As per your suggestion reverted proto file changes. Now I have pushed the code for Integration test also, please have review latest commit.

Balaji-Jatla commented:
> Can you also add S1AP integ test to avoid future regression.

Yes we already done the coding part , left with the testing.

Balaji-Jatla commented:
> I've pulled the changes and built a new mme version and it seems like the issue is still present
> 
> ```
> Jan 20 01:03:47 phy-agw1 pipelined[2673709]: Traceback (most recent call last):
> Jan 20 01:03:47 phy-agw1 pipelined[2673709]:   File "/usr/local/lib/python3.8/dist-packages/grpc/_common.py", line 86, in _transform
> Jan 20 01:03:47 phy-agw1 pipelined[2673709]:     return transformer(message)
> Jan 20 01:03:47 phy-agw1 pipelined[2673709]:   File "/usr/local/lib/python3.8/dist-packages/google/protobuf/internal/python_message.py", line 1093, in SerializeToString
> Jan 20 01:03:47 phy-agw1 pipelined[2673709]:     if not self.IsInitialized():
> Jan 20 01:03:47 phy-agw1 pipelined[2673709]: AttributeError: 'NoneType' object has no attribute 'IsInitialized'
> ```
Could you please  enable 5G feature flag as True in ```pipelined.yaml``` file, ```enable5g_features: true``` under ```/etc/magma/pipelined.yml ```. After that restart the pipelined service.

Balaji-Jatla commented:
Before executing the test
```
vagrant@magma-dev-focal:~$ cat /etc/magma/pipelined.yml |grep enable5g_features
enable5g_features: false

vagrant@magma-dev-focal:~$ cat /etc/magma/spgw.yml |grep pipelined_managed_tbl0
pipelined_managed_tbl0: false
```
executed the test case like below:
``` 
vagrant@magma-test:~/magma/lte/gateway/python/integ_tests$ make integ_test TESTS=s1aptests/test_attach_detach_attach_ul_tcp_data_tbl0.py
. /home/vagrant/build/python/bin/activate
echo "Running test: s1aptests/test_attach_detach_attach_ul_tcp_data_tbl0.py"
Running test: s1aptests/test_attach_detach_attach_ul_tcp_data_tbl0.py
timeout --foreground -k 930s 900s sudo -E PATH=/usr/lib/ccache:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/local/go/bin PYTHONPATH=:/home/vagrant/s1ap-tester/bin /home/vagrant/build/python/bin/nosetests --with-xunit --xunit-file=/var/tmp/test_results/test_attach_detach_attach_ul_tcp_data_tbl0.xml -x -s s1aptests/test_attach_detach_attach_ul_tcp_data_tbl0.py || exit 1
Start time 17:16:00
tag: AUTH_TYPE tagVal: MILENAGE
************************* Enb tester config



**********************TPT open server******** 0



Bind failed for Server {[ConID]-[FD]} ={[0]-[13]} with reason =[98]
************* Testing *Time elapsed(in usec): for timer expiry :431 usec



**********************TPT open server******** 0



Bind failed for Server {[ConID]-[FD]} ={[0]-[23]} with reason =[98]
************* Testing *Time elapsed(in usec): for timer expiry :9720 usec



**********************TPT open server******** 0



Bind failed for Server {[ConID]-[FD]} ={[0]-[23]} with reason =[98]
************* Testing *Time elapsed(in usec): for timer expiry :9832 usec



**********************TPT open server******** 0



Bind failed for Server {[ConID]-[FD]} ={[0]-[23]} with reason =[98]
************* Testing *Time elapsed(in usec): for timer expiry :9810 usec



**********************TPT open server******** 0



************* Testing *Time elapsed(in usec): for timer expiry :9951 usec
AGW is stateless
APN Correction configured
pipelined_managed_tbl0 enabled
enable5g_features enabled
Health service is disabled
Using subscriber IMSI IMSI001010000000001
Using IMEI 3805468432113171
************************* UE device config for ue_id 1
************************* Configuring IP block
************************* Waiting for IP changes to propagate
************************* S1 setup
************************* UE App config
************************* Running 1st End to End attach for UE id 1
enable5g_features enabled
pipelined_managed_tbl0 enabled
```
After executing the test
```
vagrant@magma-dev-focal:~$ cat /etc/magma/pipelined.yml |grep enable5g_features
enable5g_features: true

vagrant@magma-dev-focal:~$ cat /etc/magma/spgw.yml |grep pipelined_managed_tbl0
pipelined_managed_tbl0: true
```

koolzz commented on docs/docusaurus/sidebars.json:
```diff
@@ -274,6 +274,8 @@
         "ids": [
           "lte/dev_notes",
+          "lte/openvswitch",
+          "lte/pipelined",
```
Hm?

Balaji-Jatla replied:
It might be auto added file, reverted it. By mistake clicked on resolve conversation, sorry for that.

koolzz commented on lte/gateway/python/integ_tests/defs.mk:
```diff
@@ -167,7 +167,10 @@ s1aptests/test_guti_attach_with_zero_mtmsi.py \
 s1aptests/test_attach_implicit_detach_timer_expiry.py \
-s1aptests/test_restore_mme_config_after_sanity.py
+s1aptests/test_restore_mme_config_after_sanity.py \
+s1aptests/test_attach_detach_attach_ul_tcp_data_tbl0.py \
```
Maybe lets rename these to test_attach_detach_attach_ul_tcp_data_upf_tbl0.py

koolzz replied:
Or pipelined instead of upf

Balaji-Jatla replied:
Yeah renamed the file test_attach_detach_attach_ul_tcp_data_tbl0.py --> test_attach_detach_attach_ul_tcp_data_pipelined_tbl0.py as suggested.

koolzz reviewed (approved):
Lgtm, left minor comments. Maybe someone with more s1 tester experience can check if everything is good as well

koolzz commented on lte/gateway/c/core/oai/lib/pipelined_client/PipelinedClientAPI.cpp:
```diff
@@ -42,8 +42,14 @@ int upf_classifier_add_tunnel(
     struct ip_flow_dl* flow_dl, uint32_t flow_precedence_dl, const char* apn) {
-  std::string imsi_str = std::string(imsi);
-  std::string apn_str  = std::string(apn);
+  std::string imsi_str = "";
```
when is imsi null?

Balaji-Jatla replied:
Not sure when IMSI came as NULL, but for secure better put the null check.

koolzz replied:
It comes from MME correct? @ardzoht any idea when imsi can be null in this case? I think we should track this and investigate if this is not expected behaviour

koolzz commented:
@Balaji-Jatla  I ran integ test for this and the pipelined tbl0 managed stuff is enabled even in not pipelined managed tbl0 tests. lets fix that https://github.com/magma/magma/runs/5125876992?check_suite_focus=true#step:7:25094

f.e
```
sleep 1; 	echo "Running test: s1aptests/test_attach_restricted_plmn.py"
Running test: s1aptests/test_attach_restricted_plmn.py
timeout --foreground -k 930s 900s sudo -E PATH=/home/vagrant/build/python/bin:/home/vagrant/build/python/bin:/usr/lib/ccache:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin PYTHONPATH=:/home/vagrant/s1ap-tester/bin:/home/vagrant/s1ap-tester/bin /home/vagrant/build/python/bin/nosetests --with-xunit --xunit-file=/var/tmp/test_results/test_attach_restricted_plmn.xml -x -s s1aptests/test_attach_restricted_plmn.py || exit 1
Start time 15:51:43
tag: AUTH_TYPE tagVal: MILENAGE
************************* Enb tester config

**********************TPT open server******** 0

 Server Socket FD =[13]
************* Testing *Time elapsed(in usec): for timer expiry               :14025 usec
AGW is stateless
APN Correction configured
pipelined_managed_tbl0 enabled 
```

Also could you rebase and verify that the whole s1 integ test test suite passes fine?

Balaji-Jatla commented:
> @Balaji-Jatla I ran integ test for this and the pipelined tbl0 managed stuff is enabled even in not pipelined managed tbl0 tests. lets fix that https://github.com/magma/magma/runs/5125876992?check_suite_focus=true#step:7:25094
> 
> f.e
> 
> ```
> sleep 1; 	echo "Running test: s1aptests/test_attach_restricted_plmn.py"
> Running test: s1aptests/test_attach_restricted_plmn.py
> timeout --foreground -k 930s 900s sudo -E PATH=/home/vagrant/build/python/bin:/home/vagrant/build/python/bin:/usr/lib/ccache:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin PYTHONPATH=:/home/vagrant/s1ap-tester/bin:/home/vagrant/s1ap-tester/bin /home/vagrant/build/python/bin/nosetests --with-xunit --xunit-file=/var/tmp/test_results/test_attach_restricted_plmn.xml -x -s s1aptests/test_attach_restricted_plmn.py || exit 1
> Start time 15:51:43
> tag: AUTH_TYPE tagVal: MILENAGE
> ************************* Enb tester config
> 
> **********************TPT open server******** 0
> 
>  Server Socket FD =[13]
> ************* Testing *Time elapsed(in usec): for timer expiry               :14025 usec
> AGW is stateless
> APN Correction configured
> pipelined_managed_tbl0 enabled 
> ```
> 
> Also could you rebase and verify that the whole s1 integ test test suite passes fine?


> @Balaji-Jatla I ran integ test for this and the pipelined tbl0 managed stuff is enabled even in not pipelined managed tbl0 tests. lets fix that https://github.com/magma/magma/runs/5125876992?check_suite_focus=true#step:7:25094
> 
> f.e
> 
> ```
> sleep 1; 	echo "Running test: s1aptests/test_attach_restricted_plmn.py"
> Running test: s1aptests/test_attach_restricted_plmn.py
> timeout --foreground -k 930s 900s sudo -E PATH=/home/vagrant/build/python/bin:/home/vagrant/build/python/bin:/usr/lib/ccache:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin PYTHONPATH=:/home/vagrant/s1ap-tester/bin:/home/vagrant/s1ap-tester/bin /home/vagrant/build/python/bin/nosetests --with-xunit --xunit-file=/var/tmp/test_results/test_attach_restricted_plmn.xml -x -s s1aptests/test_attach_restricted_plmn.py || exit 1
> Start time 15:51:43
> tag: AUTH_TYPE tagVal: MILENAGE
> ************************* Enb tester config
> 
> **********************TPT open server******** 0
> 
>  Server Socket FD =[13]
> ************* Testing *Time elapsed(in usec): for timer expiry               :14025 usec
> AGW is stateless
> APN Correction configured
> pipelined_managed_tbl0 enabled 
> ```
> 
> Also could you rebase and verify that the whole s1 integ test test suite passes fine?

Ok @koolzz  Let me work on that.

Balaji-Jatla committed:
fix(mme): fixing Pipelined managed tbl0 error when mme send UE info to pipelined

Signed-off-by: Balaji-Jatla <sai.balaji@wavelabs.ai>

Balaji-Jatla commented:
> > @Balaji-Jatla I ran integ test for this and the pipelined tbl0 managed stuff is enabled even in not pipelined managed tbl0 tests. lets fix that https://github.com/magma/magma/runs/5125876992?check_suite_focus=true#step:7:25094
> > f.e
> > ```
> > sleep 1; 	echo "Running test: s1aptests/test_attach_restricted_plmn.py"
> > Running test: s1aptests/test_attach_restricted_plmn.py
> > timeout --foreground -k 930s 900s sudo -E PATH=/home/vagrant/build/python/bin:/home/vagrant/build/python/bin:/usr/lib/ccache:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin PYTHONPATH=:/home/vagrant/s1ap-tester/bin:/home/vagrant/s1ap-tester/bin /home/vagrant/build/python/bin/nosetests --with-xunit --xunit-file=/var/tmp/test_results/test_attach_restricted_plmn.xml -x -s s1aptests/test_attach_restricted_plmn.py || exit 1
> > Start time 15:51:43
> > tag: AUTH_TYPE tagVal: MILENAGE
> > ************************* Enb tester config
> > 
> > **********************TPT open server******** 0
> > 
> >  Server Socket FD =[13]
> > ************* Testing *Time elapsed(in usec): for timer expiry               :14025 usec
> > AGW is stateless
> > APN Correction configured
> > pipelined_managed_tbl0 enabled 
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Also could you rebase and verify that the whole s1 integ test test suite passes fine?
> 
> > @Balaji-Jatla I ran integ test for this and the pipelined tbl0 managed stuff is enabled even in not pipelined managed tbl0 tests. lets fix that https://github.com/magma/magma/runs/5125876992?check_suite_focus=true#step:7:25094
> > f.e
> > ```
> > sleep 1; 	echo "Running test: s1aptests/test_attach_restricted_plmn.py"
> > Running test: s1aptests/test_attach_restricted_plmn.py
> > timeout --foreground -k 930s 900s sudo -E PATH=/home/vagrant/build/python/bin:/home/vagrant/build/python/bin:/usr/lib/ccache:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin PYTHONPATH=:/home/vagrant/s1ap-tester/bin:/home/vagrant/s1ap-tester/bin /home/vagrant/build/python/bin/nosetests --with-xunit --xunit-file=/var/tmp/test_results/test_attach_restricted_plmn.xml -x -s s1aptests/test_attach_restricted_plmn.py || exit 1
> > Start time 15:51:43
> > tag: AUTH_TYPE tagVal: MILENAGE
> > ************************* Enb tester config
> > 
> > **********************TPT open server******** 0
> > 
> >  Server Socket FD =[13]
> > ************* Testing *Time elapsed(in usec): for timer expiry               :14025 usec
> > AGW is stateless
> > APN Correction configured
> > pipelined_managed_tbl0 enabled 
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Also could you rebase and verify that the whole s1 integ test test suite passes fine?
> 
> Ok @koolzz Let me work on that.

Thanks @koolzz , I found that due to below piece of code it was calling for all the test cases. Now I have removed the below piece of code and updated the PR.
<img width="509" alt="pipelined" src="https://user-images.githubusercontent.com/82372634/154085416-8839b92c-ac90-4604-b74c-95f7c4e5c3b0.PNG">

koolzz commented:
> > > @Balaji-Jatla I ran integ test for this and the pipelined tbl0 managed stuff is enabled even in not pipelined managed tbl0 tests. lets fix that https://github.com/magma/magma/runs/5125876992?check_suite_focus=true#step:7:25094
> > > f.e
> > > ```
> > > sleep 1; 	echo "Running test: s1aptests/test_attach_restricted_plmn.py"
> > > Running test: s1aptests/test_attach_restricted_plmn.py
> > > timeout --foreground -k 930s 900s sudo -E PATH=/home/vagrant/build/python/bin:/home/vagrant/build/python/bin:/usr/lib/ccache:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin PYTHONPATH=:/home/vagrant/s1ap-tester/bin:/home/vagrant/s1ap-tester/bin /home/vagrant/build/python/bin/nosetests --with-xunit --xunit-file=/var/tmp/test_results/test_attach_restricted_plmn.xml -x -s s1aptests/test_attach_restricted_plmn.py || exit 1
> > > Start time 15:51:43
> > > tag: AUTH_TYPE tagVal: MILENAGE
> > > ************************* Enb tester config
> > > 
> > > **********************TPT open server******** 0
> > > 
> > >  Server Socket FD =[13]
> > > ************* Testing *Time elapsed(in usec): for timer expiry               :14025 usec
> > > AGW is stateless
> > > APN Correction configured
> > > pipelined_managed_tbl0 enabled 
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Also could you rebase and verify that the whole s1 integ test test suite passes fine?
> > 
> > 
> > > @Balaji-Jatla I ran integ test for this and the pipelined tbl0 managed stuff is enabled even in not pipelined managed tbl0 tests. lets fix that https://github.com/magma/magma/runs/5125876992?check_suite_focus=true#step:7:25094
> > > f.e
> > > ```
> > > sleep 1; 	echo "Running test: s1aptests/test_attach_restricted_plmn.py"
> > > Running test: s1aptests/test_attach_restricted_plmn.py
> > > timeout --foreground -k 930s 900s sudo -E PATH=/home/vagrant/build/python/bin:/home/vagrant/build/python/bin:/usr/lib/ccache:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin PYTHONPATH=:/home/vagrant/s1ap-tester/bin:/home/vagrant/s1ap-tester/bin /home/vagrant/build/python/bin/nosetests --with-xunit --xunit-file=/var/tmp/test_results/test_attach_restricted_plmn.xml -x -s s1aptests/test_attach_restricted_plmn.py || exit 1
> > > Start time 15:51:43
> > > tag: AUTH_TYPE tagVal: MILENAGE
> > > ************************* Enb tester config
> > > 
> > > **********************TPT open server******** 0
> > > 
> > >  Server Socket FD =[13]
> > > ************* Testing *Time elapsed(in usec): for timer expiry               :14025 usec
> > > AGW is stateless
> > > APN Correction configured
> > > pipelined_managed_tbl0 enabled 
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Also could you rebase and verify that the whole s1 integ test test suite passes fine?
> > 
> > 
> > Ok @koolzz Let me work on that.
> 
> Thanks @koolzz , I found that due to below piece of code it was calling for all the test cases. Now I have removed the below piece of code and updated the PR. <img alt="pipelined" width="509" src="https://user-images.githubusercontent.com/82372634/154085416-8839b92c-ac90-4604-b74c-95f7c4e5c3b0.PNG">

Thanks, and have you verified that the s1 test suite completes successfully?

koolzz commented:
@Balaji-Jatla Could you rebase, I'm getting this
```
 sleep 1;        echo "Running test: s1aptests/test_attach_detach_attach_ul_tcp_data_pipelined_tbl0.py"
Running test: s1aptests/test_attach_detach_attach_ul_tcp_data_pipelined_tbl0.py
timeout --foreground -k 930s 900s sudo -E PATH=/usr/lib/ccache:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/local/go/bin PYTHONPATH=:/home/vagrant/s1ap-tester/bin /home/vagrant/build/python/bin/nosetests --with-xunit --xunit-file=/var/tmp/test_results/test_attach_detach_attach_ul_tcp_data_pipelined_tbl0.xml -x -s s1aptests/test_attach_detach_attach_ul_tcp_data_pipelined_tbl0.py || exit 1
Start time 21:56:33
tag: AUTH_TYPE tagVal: MILENAGE
************************* Enb tester config

**********************TPT open server******** 0


 Server Socket FD =[13]
************* Testing *Time elapsed(in usec): for timer expiry               :6931 usec
AGW is stateless
APN Correction configured
Health service is disabled
Using subscriber IMSI IMSI001010000000001
Using IMEI 3805468432113171
************************* UE device config for ue_id  1
************************* Configuring IP block
************************* Waiting for IP changes to propagate
************************* S1 setup
************************* UE App config
************************* Running 1st End to End attach for  UE id  1
enable5g_features enabled
pipelined_managed_tbl0 enabled

************************* send SCTP SHUTDOWN
Keys left in Redis (list should be empty)[
  
]
Entries in s1ap_imsi_map (should be zero): 0
Entries left in hashtables (should be zero): 0
Entries in mme_ueip_imsi_map (should be zero): 0
F
======================================================================
FAIL: Attach, detach, re-attach, send UL TCP data, and detach
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/vagrant/magma/lte/gateway/python/integ_tests/s1aptests/test_attach_detach_attach_ul_tcp_data_pipelined_tbl0.py", line 54, in test_attach_detach_attach_ul_tcp_data_pipelined_tbl0
    s1ap_types.ueAttachAccept_t,
  File "/home/vagrant/magma/lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py", line 375, in attach
    assert resp_type.value == response.msg_type
AssertionError: 
-------------------- >> begin captured logging << --------------------
root: ERROR: Could not load magmad yml config for mconfig modules
urllib3.connectionpool: DEBUG: Starting new HTTP connection (1): 192.168.60.142:3333
urllib3.connectionpool: DEBUG: http://192.168.60.142:3333 "GET /stats/switches HTTP/1.1" 200 17
root: INFO: Adding subscriber : IMSI001010000000001
root: DEBUG: s1ap response expected, received: 12, 14
root: INFO: Deleting subscriber : IMSI001010000000001
--------------------- >> end captured logging << ---------------------

----------------------------------------------------------------------
Ran 1 test in 214.839s

FAILED (failures=1)
Makefile:42: recipe for target 'integ_test' failed
make: *** [integ_test] Error 1
vagrant@magma-test:~/magma/lte/gateway/python/integ_tests$ 

```

Balaji-Jatla commented:
> @Balaji-Jatla Could you rebase, I'm getting this
> 
> ```
>  sleep 1;        echo "Running test: s1aptests/test_attach_detach_attach_ul_tcp_data_pipelined_tbl0.py"
> Running test: s1aptests/test_attach_detach_attach_ul_tcp_data_pipelined_tbl0.py
> timeout --foreground -k 930s 900s sudo -E PATH=/usr/lib/ccache:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/local/go/bin PYTHONPATH=:/home/vagrant/s1ap-tester/bin /home/vagrant/build/python/bin/nosetests --with-xunit --xunit-file=/var/tmp/test_results/test_attach_detach_attach_ul_tcp_data_pipelined_tbl0.xml -x -s s1aptests/test_attach_detach_attach_ul_tcp_data_pipelined_tbl0.py || exit 1
> Start time 21:56:33
> tag: AUTH_TYPE tagVal: MILENAGE
> ************************* Enb tester config
> 
> **********************TPT open server******** 0
> 
> 
>  Server Socket FD =[13]
> ************* Testing *Time elapsed(in usec): for timer expiry               :6931 usec
> AGW is stateless
> APN Correction configured
> Health service is disabled
> Using subscriber IMSI IMSI001010000000001
> Using IMEI 3805468432113171
> ************************* UE device config for ue_id  1
> ************************* Configuring IP block
> ************************* Waiting for IP changes to propagate
> ************************* S1 setup
> ************************* UE App config
> ************************* Running 1st End to End attach for  UE id  1
> enable5g_features enabled
> pipelined_managed_tbl0 enabled
> 
> ************************* send SCTP SHUTDOWN
> Keys left in Redis (list should be empty)[
>   
> ]
> Entries in s1ap_imsi_map (should be zero): 0
> Entries left in hashtables (should be zero): 0
> Entries in mme_ueip_imsi_map (should be zero): 0
> F
> ======================================================================
> FAIL: Attach, detach, re-attach, send UL TCP data, and detach
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/home/vagrant/magma/lte/gateway/python/integ_tests/s1aptests/test_attach_detach_attach_ul_tcp_data_pipelined_tbl0.py", line 54, in test_attach_detach_attach_ul_tcp_data_pipelined_tbl0
>     s1ap_types.ueAttachAccept_t,
>   File "/home/vagrant/magma/lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py", line 375, in attach
>     assert resp_type.value == response.msg_type
> AssertionError: 
> -------------------- >> begin captured logging << --------------------
> root: ERROR: Could not load magmad yml config for mconfig modules
> urllib3.connectionpool: DEBUG: Starting new HTTP connection (1): 192.168.60.142:3333
> urllib3.connectionpool: DEBUG: http://192.168.60.142:3333 "GET /stats/switches HTTP/1.1" 200 17
> root: INFO: Adding subscriber : IMSI001010000000001
> root: DEBUG: s1ap response expected, received: 12, 14
> root: INFO: Deleting subscriber : IMSI001010000000001
> --------------------- >> end captured logging << ---------------------
> 
> ----------------------------------------------------------------------
> Ran 1 test in 214.839s
> 
> FAILED (failures=1)
> Makefile:42: recipe for target 'integ_test' failed
> make: *** [integ_test] Error 1
> vagrant@magma-test:~/magma/lte/gateway/python/integ_tests$ 
> ```

@koolzz I have rebased the PR , but still unable to test it. Even though we reported in multiple channels in slack couldn't get proper solution yet. Could you please help in this issue.
![image](https://user-images.githubusercontent.com/82372634/154729547-c7140c21-bf2b-4f21-adba-f1a120c75650.png)

koolzz commented:
> 

The failing test is the one you added, can you verify datapath flows are installed?

VinashakAnkitAman commented on lte/gateway/python/integ_tests/s1aptests/test_attach_detach_attach_dl_tcp_data_pipelined_tbl0.py:
```diff
@@ -0,0 +1,112 @@
+                MagmadUtil.pipelined_managed_tbl0_cmds.ENABLE,
+        )
+        # To reflect above config restart pipelined service is mandatory
+        self._s1ap_wrapper.magmad_util.restart_services(["pipelined"])
```
Please add 30 sec sleep after every service restart in this PR change. Sample code is in testcase: s1aptests/test_attach_detach_with_mme_restart.py

GANESH-WAVELABS replied:
Thanks for letting us know this after adding this wait time test cases become success.

GANESH-WAVELABS commented:
@koolzz and @pshelar We have added some wait time in new test cases after restarting the pipelined service, now test cases became success. Please find the screen shots and logs below:
test case screen logs in sequential manner:
![test_case_table0_ul_attach_detach_1](https://user-images.githubusercontent.com/70899516/155736901-74b9a79c-9ac1-47b7-b8a3-6892100077ad.JPG)
![test_case_table0_ul_attach_detach_2](https://user-images.githubusercontent.com/70899516/155736940-30c3b919-d892-40ff-a1d0-f4d1b4a90f6d.JPG)
![test_case_table0_ul_attach_detach_3](https://user-images.githubusercontent.com/70899516/155736975-8dcf25cc-902f-4be8-ad99-72ade611ac6a.JPG)
![test_case_table0_ul_attach_detach_4](https://user-images.githubusercontent.com/70899516/155737000-591aa18a-fbc0-4c7d-bc2f-a51751e4727b.JPG)
![test_case_table0_ul_attach_detach_5](https://user-images.githubusercontent.com/70899516/155737014-a0a0aa8b-14c7-4aec-8c5b-1308cbadb669.JPG)
Below screen shot shows the `pipelined_managed_tbl0` flag status from `spgw.yml` before and after executing the test case:
![pipelined_managed_table0_falg_from_spgw_yml](https://user-images.githubusercontent.com/70899516/155737481-cf5b8f5b-21de-4476-9f21-bfee43fd13e8.JPG)

Below screen shot shows the `enable5g_features` flag status from `pipelined.yml` before and after executing the test case:
![enable5g_features_flag_from_pipeliend_yml](https://user-images.githubusercontent.com/70899516/155737575-ee2f9335-5372-45e9-bd43-76e9e903775f.JPG)

OVS table0 entries during UL attach detach test case:
![table0_entries_during_ul_attach_detach](https://user-images.githubusercontent.com/70899516/155737770-2dcb3bc5-5eb3-4d71-a987-a25b5efedadf.JPG)

OVS table0 entries during DL attach detach test case:
![table0_entries_during_dl_attach_detach](https://user-images.githubusercontent.com/70899516/155737810-b4bab3bb-7101-4ad2-9171-ceb4cf96dc93.JPG)

GANESH-WAVELABS commented:
Screen logs during UL attach detach test case:
```
(python) vagrant@magma-test:~/magma/lte/gateway/python/integ_tests$ make integ_test TESTS=s1aptests/test_attach_detach_attach_ul_tcp_data_pipelined_tbl0.py
. /home/vagrant/build/python/bin/activate
echo "Running test: s1aptests/test_attach_detach_attach_ul_tcp_data_pipelined_tbl0.py"
Running test: s1aptests/test_attach_detach_attach_ul_tcp_data_pipelined_tbl0.py
timeout --foreground -k 930s 900s sudo -E PATH=/home/vagrant/build/python/bin:/usr/lib/ccache:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/local/go/bin PYTHONPATH=:/home/vagrant/s1ap-tester/bin:/home/vagrant/s1ap-tester/bin /home/vagrant/build/python/bin/nosetests --with-xunit --xunit-file=/var/tmp/test_results/test_attach_detach_attach_ul_tcp_data_pipelined_tbl0.xml -x -s s1aptests/test_attach_detach_attach_ul_tcp_data_pipelined_tbl0.py || exit 1
Start time 12:01:34
tag: AUTH_TYPE tagVal: MILENAGE
************************* Enb tester config

**********************TPT open server******** 0

 Server Socket FD =[13]
************* Testing *Time elapsed(in usec): for timer expiry               :445 usec
AGW is stateless
APN Correction configured
Health service is disabled
Using subscriber IMSI IMSI001010000000001
Using IMEI 3805468432113171
************************* UE device config for ue_id  1
************************* Configuring IP block
************************* Waiting for IP changes to propagate
************************* S1 setup
************************* UE App config
************************* Running 1st End to End attach for  UE id  1
enable5g_features enabled
pipelined_managed_tbl0 enabled
Waiting for 0 seconds
Waiting for 1 seconds
Waiting for 2 seconds
Waiting for 3 seconds
Waiting for 4 seconds
Waiting for 5 seconds
Waiting for 6 seconds
Waiting for 7 seconds
Waiting for 8 seconds
Waiting for 9 seconds
Waiting for 10 seconds
Waiting for 11 seconds
Waiting for 12 seconds
Waiting for 13 seconds
Waiting for 14 seconds
Waiting for 15 seconds
Waiting for 16 seconds
Waiting for 17 seconds
Waiting for 18 seconds
Waiting for 19 seconds
Waiting for 20 seconds
Waiting for 21 seconds
Waiting for 22 seconds
Waiting for 23 seconds
Waiting for 24 seconds
Waiting for 25 seconds
Waiting for 26 seconds
Waiting for 27 seconds
Waiting for 28 seconds
Waiting for 29 seconds
ue id 1 found

*************NbUiNbuHdlUeRadioCapMsg: Send UE cap msg SUCCESS*********
PDN_TYPE: 1
IPv4 PDN_ADDRESS:192 168 128 12
************************* Running UE detach for UE id  1
Deleting Ue Context from Traffic Handler
************************* Running 2nd End to End attach for  UE id  1
ue id 1 found

*************NbUiNbuHdlUeRadioCapMsg: Send UE cap msg SUCCESS*********
PDN_TYPE: 1
IPv4 PDN_ADDRESS:192 168 128 12
************************* Running UE uplink (TCP) for UE id  1
Connecting to host 192.168.129.42, port 5001
[ 29] local 192.168.128.12 port 39345 connected to 192.168.129.42 port 5001
[ ID] Interval           Transfer     Bandwidth       Retr  Cwnd
[ 29]   0.00-1.00   sec  18.6 KBytes   152 Kbits/sec    0   5.62 KBytes
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bandwidth       Retr
[ 29]   0.00-1.00   sec  18.6 KBytes   152 Kbits/sec    0             sender
[ 29]   0.00-1.00   sec  16.9 KBytes   138 Kbits/sec                  receiver

iperf Done.
************************* Running UE detach for UE id  1
Deleting Ue Context from Traffic Handler
enable5g_features disabled
pipelined_managed_tbl0 disabled
Waiting for 0 seconds
Waiting for 1 seconds
Waiting for 2 seconds
Waiting for 3 seconds
Waiting for 4 seconds
Waiting for 5 seconds
Waiting for 6 seconds
Waiting for 7 seconds
Waiting for 8 seconds
Waiting for 9 seconds
Waiting for 10 seconds
Waiting for 11 seconds
Waiting for 12 seconds
Waiting for 13 seconds
Waiting for 14 seconds
Waiting for 15 seconds
Waiting for 16 seconds
Waiting for 17 seconds
Waiting for 18 seconds
Waiting for 19 seconds
Waiting for 20 seconds
Waiting for 21 seconds
Waiting for 22 seconds
Waiting for 23 seconds
Waiting for 24 seconds
Waiting for 25 seconds
Waiting for 26 seconds
Waiting for 27 seconds
Waiting for 28 seconds
Waiting for 29 seconds
************************* send SCTP SHUTDOWN
Keys left in Redis (list should be empty)[

]
Entries in s1ap_imsi_map (should be zero): 0
Entries left in hashtables (should be zero): 0
Entries in mme_ueip_imsi_map (should be zero): 0
.
----------------------------------------------------------------------
Ran 1 test in 89.265s

OK
sleep 1
(python) vagrant@magma-test:~/magma/lte/gateway/python/integ_tests$
```
Screen logs during DL attach detach test case:
```
(python) vagrant@magma-test:~/magma/lte/gateway/python/integ_tests$ make integ_test TESTS=s1aptests/test_attach_detach_attach_dl_tcp_data_pipelined_tbl0.py
. /home/vagrant/build/python/bin/activate
echo "Running test: s1aptests/test_attach_detach_attach_dl_tcp_data_pipelined_tbl0.py"
Running test: s1aptests/test_attach_detach_attach_dl_tcp_data_pipelined_tbl0.py
timeout --foreground -k 930s 900s sudo -E PATH=/home/vagrant/build/python/bin:/usr/lib/ccache:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/local/go/bin PYTHONPATH=:/home/vagrant/s1ap-tester/bin:/home/vagrant/s1ap-tester/bin /home/vagrant/build/python/bin/nosetests --with-xunit --xunit-file=/var/tmp/test_results/test_attach_detach_attach_dl_tcp_data_pipelined_tbl0.xml -x -s s1aptests/test_attach_detach_attach_dl_tcp_data_pipelined_tbl0.py || exit 1
Start time 12:03:46
tag: AUTH_TYPE tagVal: MILENAGE
************************* Enb tester config

**********************TPT open server******** 0

 Server Socket FD =[13]
************* Testing *Time elapsed(in usec): for timer expiry               :6956 usec
AGW is stateless
APN Correction configured
Health service is disabled
Using subscriber IMSI IMSI001010000000001
Using IMEI 3805468432113171
************************* UE device config for ue_id  1
************************* Configuring IP block
************************* Waiting for IP changes to propagate
************************* S1 setup
************************* UE App config
************************* Running 1st End to End attach for  UE id  1
enable5g_features enabled
pipelined_managed_tbl0 enabled
Waiting for 0 seconds
Waiting for 1 seconds
Waiting for 2 seconds
Waiting for 3 seconds
Waiting for 4 seconds
Waiting for 5 seconds
Waiting for 6 seconds
Waiting for 7 seconds
Waiting for 8 seconds
Waiting for 9 seconds
Waiting for 10 seconds
Waiting for 11 seconds
Waiting for 12 seconds
Waiting for 13 seconds
Waiting for 14 seconds
Waiting for 15 seconds
Waiting for 16 seconds
Waiting for 17 seconds
Waiting for 18 seconds
Waiting for 19 seconds
Waiting for 20 seconds
Waiting for 21 seconds
Waiting for 22 seconds
Waiting for 23 seconds
Waiting for 24 seconds
Waiting for 25 seconds
Waiting for 26 seconds
Waiting for 27 seconds
Waiting for 28 seconds
Waiting for 29 seconds
ue id 1 found

*************NbUiNbuHdlUeRadioCapMsg: Send UE cap msg SUCCESS*********
PDN_TYPE: 1
IPv4 PDN_ADDRESS:192 168 128 12
************************* Running UE detach for UE id  1
Deleting Ue Context from Traffic Handler
************************* Running 2nd End to End attach for  UE id  1
ue id 1 found

*************NbUiNbuHdlUeRadioCapMsg: Send UE cap msg SUCCESS*********
PDN_TYPE: 1
IPv4 PDN_ADDRESS:192 168 128 12
************************* Running UE downlink (TCP) for UE id  1
-----------------------------------------------------------
Server listening on 7001
-----------------------------------------------------------
Accepted connection from 192.168.129.42, port 34553
[ 30] local 192.168.128.12 port 7001 connected to 192.168.129.42 port 42941
[ ID] Interval           Transfer     Bandwidth
[ 30]   0.00-1.00   sec  1.17 MBytes  9.83 Mbits/sec
[ 30]   1.00-1.05   sec  0.00 Bytes  0.00 bits/sec
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bandwidth
[ 30]   0.00-1.05   sec  0.00 Bytes  0.00 bits/sec                  sender
[ 30]   0.00-1.05   sec  1.17 MBytes  9.35 Mbits/sec                  receiver
************************* Running UE detach for UE id  1
Deleting Ue Context from Traffic Handler
enable5g_features disabled
pipelined_managed_tbl0 disabled
Waiting for 0 seconds
Waiting for 1 seconds
Waiting for 2 seconds
Waiting for 3 seconds
Waiting for 4 seconds
Waiting for 5 seconds
Waiting for 6 seconds
Waiting for 7 seconds
Waiting for 8 seconds
Waiting for 9 seconds
Waiting for 10 seconds
Waiting for 11 seconds
Waiting for 12 seconds
Waiting for 13 seconds
Waiting for 14 seconds
Waiting for 15 seconds
Waiting for 16 seconds
Waiting for 17 seconds
Waiting for 18 seconds
Waiting for 19 seconds
Waiting for 20 seconds
Waiting for 21 seconds
Waiting for 22 seconds
Waiting for 23 seconds
Waiting for 24 seconds
Waiting for 25 seconds
Waiting for 26 seconds
Waiting for 27 seconds
Waiting for 28 seconds
Waiting for 29 seconds
************************* send SCTP SHUTDOWN
Keys left in Redis (list should be empty)[

]
Entries in s1ap_imsi_map (should be zero): 0
Entries left in hashtables (should be zero): 0
Entries in mme_ueip_imsi_map (should be zero): 0
.
----------------------------------------------------------------------
Ran 1 test in 89.234s

OK
sleep 1
(python) vagrant@magma-test:~/magma/lte/gateway/python/integ_tests$
```
OVS table0 entries during UL attach detach test case:
```
Every 2.0s: sudo ovs-ofctl -O OpenFlow13 dump-flows gtp_br0 table=0 magma-dev-focal: Fri Feb 25 14:30:26 2022

OFPST_FLOW reply (OF1.3) (xid=0x6):
 cookie=0x0, duration=32.822s, table=0, n_packets=11, n_bytes=770, priority=10,in_port=15578 actions=goto_tab
le:201
 cookie=0x0, duration=27.194s, table=0, n_packets=0, n_bytes=0, priority=0,in_port=15578 actions=resubmit(,20
1)
 cookie=0x0, duration=27.194s, table=0, n_packets=9, n_bytes=630, priority=0,in_port=15579 actions=resubmit(,
202)
 cookie=0x0, duration=10.261s, table=0, n_packets=50, n_bytes=20858, priority=3,tun_id=0x19,in_port=32768 act
ions=set_field:02:00:00:00:00:01->eth_src,set_field:ff:ff:ff:ff:ff:ff->eth_dst,set_field:0xa000128->reg9,set_
field:0x7594587a00d->metadata,resubmit(,1)
 cookie=0x0, duration=10.261s, table=0, n_packets=52, n_bytes=3644, priority=3,ip,in_port=LOCAL,nw_dst=192.16
8.128.12 actions=set_field:0xa000128->tun_id,set_field:192.168.60.141->tun_dst,set_field:0x8000->reg8,set_fie
ld:0x19->reg9,set_field:0x7594587a00d->metadata,resubmit(,1)
 cookie=0x0, duration=10.261s, table=0, n_packets=0, n_bytes=0, priority=3,ip,in_port=15577,nw_dst=192.168.12
8.12 actions=set_field:0xa000128->tun_id,set_field:192.168.60.141->tun_dst,set_field:0x8000->reg8,set_field:0
x19->reg9,set_field:0x7594587a00d->metadata,resubmit(,1)
 cookie=0x0, duration=10.261s, table=0, n_packets=1, n_bytes=42, priority=3,arp,in_port=LOCAL,arp_tpa=192.168
.128.12 actions=set_field:0x7594587a00d->metadata,resubmit(,1)
 cookie=0x0, duration=10.261s, table=0, n_packets=0, n_bytes=0, priority=3,arp,in_port=15577,arp_tpa=192.168.
128.12 actions=set_field:0x7594587a00d->metadata,resubmit(,1)
 cookie=0x0, duration=27.194s, table=0, n_packets=1053, n_bytes=69152, priority=0 actions=resubmit(,1)
```
OVS table0 entries during DL attach detach test case:
```
Every 2.0s: sudo ovs-ofctl -O OpenFlow13 dump-flows gtp_br0 table=0 magma-dev-focal: Fri Feb 25 14:26:09 2022

OFPST_FLOW reply (OF1.3) (xid=0x6):
 cookie=0x0, duration=39.049s, table=0, n_packets=11, n_bytes=770, priority=10,in_port=15578 actions=goto_tab
le:201
 cookie=0x0, duration=28.857s, table=0, n_packets=0, n_bytes=0, priority=0,in_port=15578 actions=resubmit(,20
1)
 cookie=0x0, duration=28.857s, table=0, n_packets=9, n_bytes=630, priority=0,in_port=15579 actions=resubmit(,
202)
 cookie=0x0, duration=12.208s, table=0, n_packets=0, n_bytes=0, priority=3,tun_id=0x16,in_port=32768 actions=
set_field:02:00:00:00:00:01->eth_src,set_field:ff:ff:ff:ff:ff:ff->eth_dst,set_field:0xa000128->reg9,set_field
:0x7594587a00d->metadata,resubmit(,1)
 cookie=0x0, duration=12.208s, table=0, n_packets=0, n_bytes=0, priority=3,ip,in_port=LOCAL,nw_dst=192.168.12
8.12 actions=set_field:0xa000128->tun_id,set_field:192.168.60.141->tun_dst,set_field:0x8000->reg8,set_field:0
x16->reg9,set_field:0x7594587a00d->metadata,resubmit(,1)
 cookie=0x0, duration=12.208s, table=0, n_packets=0, n_bytes=0, priority=3,ip,in_port=15577,nw_dst=192.168.12
8.12 actions=set_field:0xa000128->tun_id,set_field:192.168.60.141->tun_dst,set_field:0x8000->reg8,set_field:0
x16->reg9,set_field:0x7594587a00d->metadata,resubmit(,1)
 cookie=0x0, duration=12.208s, table=0, n_packets=0, n_bytes=0, priority=3,arp,in_port=LOCAL,arp_tpa=192.168.
128.12 actions=set_field:0x7594587a00d->metadata,resubmit(,1)
 cookie=0x0, duration=12.208s, table=0, n_packets=0, n_bytes=0, priority=3,arp,in_port=15577,arp_tpa=192.168.
128.12 actions=set_field:0x7594587a00d->metadata,resubmit(,1)
 cookie=0x0, duration=28.857s, table=0, n_packets=827, n_bytes=65159, priority=0 actions=resubmit(,1)
```

pshelar commented on lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py:
```diff
@@ -1015,6 +1017,65 @@ def config_apn_correction(self, cmd):
+            if ret_code == 0:
+                print("enable5g_features enabled")
+        else:
+            enable5g_features_cmd = "sed -i \'s/enable5g_features: true/enable5g_features: false/g\' /etc/magma/pipelined.yml"
```
Is there any downside of enabling enable5g_features and pipelined_managed_tbl0  by default?

prabinakpattnaik replied:
@pshelar , as such there is no impact if we enabled the both flags by default. But by default MME is using fluid for GTP tunnel setup and  as per design now, enable5g_features flag is only configured by Orc8r.  We removed the enable5g_features flag for all modules by default.

pshelar replied:
ok, lets enable enable5g_features for pipelined.
We have to keep pipelined_managed_tbl0 off by default for perf reason.

prabinakpattnaik replied:
@pshelar , Sorry for late response.. As Ulas suggested we have to make for generalizing the behavior for enable5g_features flags in yaml files and By default orc8r should drive the behavior. So that we had Commented the enable5g_features flag from all yaml files recently. (#11797 )

pshelar replied:
You you enable it for magma dev-vm pipelineD config only? keep prod config as it is.
enable it here: https://github.com/magma/magma/blob/master/lte/gateway/configs/pipelined.yml

cc: @ulaskozat

GANESH-WAVELABS replied:
@pshelar We are configuring `enable5g_features` flag in side dev-vm under `/etc/magma/pipelined.yml` file which is always sync with `magma/lte/gateway/configs/pipelined.yml`. All S1AP test cases are following same procedure like editing yml files under dev-vm's `/etc/magma/` path.  Though we have `/etc/magma/pipelined.yml_prod` there is no such  `enable5g_features` flag under `/etc/magma/pipelined.yml_prod`

pshelar replied:
I see that.
I am suggesting to set 'enable5g_features to true' in 'magma/lte/gateway/configs/pipelined.yml' to avoid chaning it in test and restart of pipelineD service in integ tests.

GANESH-WAVELABS replied:
@pshelar for changing `enable5g_features` default value to `true` in `magma/lte/gateway/configs/pipelined.yml` we need to take grant from Ulas before making changes.
CC: @ulaskozat Could you please help here to move further on this task.

ulaskozat replied:
it is fine. @pshelar please add documentation for pipelined.yml vs. pipelined.yml_prod.

pshelar replied:
sure.

GANESH-WAVELABS replied:
Sorry for the delay there was some confusion, now edited `enable5g_features` default value to `true` in `magma/lte/gateway/configs/pipelined.yml` as suggested by @pshelar and @ulaskozat .
@pshelar could you please have one more round of review and move this PR to further

pshelar replied:
@GANESH-WAVELABS some of mandatory checks are failing, can you fix it.

GANESH-WAVELABS replied:
Thanks @pshelar , I fixed the mandatory checks which was failing earlier. PR is `ready2merge` now.
